OpenRA.Mods.Common/FileFormats/ImaAdpcmReader.cs

```diff
@@ -85,5 +99,114 @@ public static byte[] LoadImaAdpcmSound(ReadOnlySpan<byte> raw, ref int index, re
 
 			return output;
 		}
+
+		public static byte[] ReadData(Stream stream, long dataOffset, int dataSize, short blockAlign, short channels)
+		{
+			const int SamplesPerGroup = 8;
+
+			ArgumentNullException.ThrowIfNull(stream);
+
+			var sourceData = SegmentStream.GetReadableData(stream, dataOffset, dataSize);
+
+			var numBlocks = dataSize / blockAlign;
+
+			var predictorSize = 4 * channels;
+			var blockDataSize = blockAlign - predictorSize;
+
+			// We get two samples per nibble
+			var samplesPerChannel = blockDataSize * 2 / channels;
+
+			// 8 samples from a 4-byte group
+			var groupCount = samplesPerChannel / SamplesPerGroup;
+
+			var predOut = numBlocks * channels * 2;
+			var groupOut = numBlocks * groupCount * channels * SamplesPerGroup * 2;
+			var estimatedOutDataSize = predOut + groupOut;
+
+			var outData = new byte[estimatedOutDataSize];
+
+			// PERF: The output is 16-bit PCM, so we can write bytes as if they were shorts for less CPU churn.
+			var outShorts = MemoryMarshal.Cast<byte, short>(outData.AsSpan());
+
+			// NOTE: decoding a block is sequentually dependent on predictor/index.
+			Span<short> predictor = stackalloc short[channels];
+			Span<byte> index = stackalloc byte[channels];
+
+			// PERF: Avoid bounds checks by using refs.
+			ref var srcRef = ref MemoryMarshal.GetReference(sourceData);
+			ref var outRef = ref MemoryMarshal.GetReference(outShorts);
+			ref var predRef = ref MemoryMarshal.GetReference(predictor);
+			ref var idxRef = ref MemoryMarshal.GetReference(index);
+
+			// Global decoded sample counter
+			var src = 0;
+			var outSample = 0;
+
+			for (var block = 0; block < numBlocks; block++)
+			{
+				// Initial states
+				for (var c = 0; c < channels; c++)
+				{
+					var offset = src + c * 4;
+
+					// Load initial values.
+					var pred = (short)(Unsafe.Add(ref srcRef, offset) | (Unsafe.Add(ref srcRef, offset + 1) << 8));
+					var idx = Unsafe.Add(ref srcRef, offset + 2);
+
+					Unsafe.Add(ref predRef, c) = pred;
+					Unsafe.Add(ref idxRef, c) = idx;
+				}
+
+				src += predictorSize;
+
+				// Write initial predictor samples interleaved
+				for (var c = 0; c < channels; c++)
+					Unsafe.Add(ref outRef, outSample + c) = Unsafe.Add(ref predRef, c);
+
+				outSample += channels;
+
+				for (var iter = 0; iter < groupCount; iter++)
```

Comment:
We're decoding in a batch of 8 at a time, I assume as a form of loop unrolling?

Is there a guarentee that chunks have a multiple of 8 in them? If not, we're need another loop to catch the leftovers or we're going to miss out some samples.

Reply:
it is guaranteed, if there's an incomplete group at the end of the file we should just skip it

Reply:
though, the fast path will just fail and decode garbage